accepted/0017-add-funding-support.md

```diff
@@ -0,0 +1,117 @@
+# Add `funding` support to `package.json`
+
+## Summary
+
+This RFC identifies an initial means of tooling to describe & notify consumers of a package's monetary support information.
+
+## Motivation
+
+Package maintainers want to clearly indentify how their software is currently, or could be in the future, supported monetarily.
+
+## Detailed Explanation
+
+* See [Prior Art](#prior-art)
+* Provide a means to **reference** existing backers or funding opportunities in `package.json`
+* Provide a means to **define the type** of backing or funding opportunities in `package.json`
+* Provide a means to **notify** package consumers of backers or funding opportunites with the cli
+* Provide a means to **view** dependency's backers or funding opportunites with the cli
+* Provide a means to **view** a package's backers or funding opportunites
+
+##  Rationale & Alternatives
+* Rationale: This is a straightforward and easily implemented solution to take the first step toward supporting Open Source Package Maintainence sustainability
+* Alternative: Do nothing
+* Alternative: Continue debating a larger, more complex schema /w correspondingly more complex tooling
+
+## Implementation
+
+* Add a `funding` field to `package.json`
+  * supports a string URL or object with...
+  * keys that support arbitrary string identifiers
+    * ex. `"sponsor"`, `"sponsors"`, `"donations"`, `"bounties"`, `"contributors"`, `"patrons"` etc.
```

Comment:
Do these string identifiers have a defined meaning? If not, then how does the CLI know which URL or URLs to display when running `npm fund`?

If the goal is to just do something quick, simple, and more useful than nothing, I'd recommend just starting with a string value that points to a URL with more information on funding, and drop support for an object with arbitrary keys. Then `npm fund` can just straightforwardly print out the funding URL for each package.

Reply:
@feross I made a few changes to this RFC based on both @djsauble's & your feedback. I'm totally okay with de-scoping the work to only focus on a singular URL although I think the hope was to pull in some of the ideas behind the PM WG's more verbose [`backing` schema](https://github.com/nodejs/package-maintenance/blob/master/docs/drafts/PACKAGE-SUPPORT.md#support-backing).

Hopefully the updated proposal for `npm fund <pkg>` makes it a bit more clear in terms of when/how we'd try to open a URL **versus** print out the information. 

Again, if we want to simplify this to only accommodate the first scenario, I'm all for that (as it would also make the `npmjs.com` changes that much simpler as well).

Reply:
One concern I have is that allowing a free-for-all with respect to these identifiers will make it harder to have consistency later on. We could perhaps come up with a set of "reserved" identifiers that people are welcome to use now with the understanding that they may have specific tooling behavior applied to them later.

But I also wouldn't mind dropping the arbitrary keys from v1 if the WG agrees.

Reply:
>  We could perhaps come up with a set of "reserved" identifiers that people are welcome to use now with the understanding that they may have specific tooling behavior applied to them later.

I think this could become problematic.  It isn't always easy to take something away from users even if we are stating it ahead of time

Reply:
I would suggest - even if it’s more boilerplate - something like:
```json
"funding": {
  "bikeshed": {
  }
}
```
where every key of “bikeshed” is arbitrary and unrestricted, and thus serves as an ID for other things - tools, the “support” field the package management group is working on - to use to address the funding mechanism.

This keeps the design space open for any future additions to the funding object, at the (obviateable vía tooling) cost of adding an additional nesting to the funding object.

Reply:
I would suggest that, at the outset, we specify that there are exactly two fields: `url` and `type`, mirroring the `repository` field we have today.  `url` is mandatory, `type` is optional, and can often be inferred from the url.  (Eg, if it's at `patreon.com`, we know it's `"type": "patreon"` and can interpret it that way in `read-package-json`.)

The spec should state that additional fields MAY be added in the future, and so it is not safe to add them arbitrarily.  People still will, but then they have fair warning when we assign an official meaning to some of them in the future.